docs/src/references/mqtt-topic-guidelines.md

```diff
@@ -0,0 +1,214 @@
+
+### Nice-to have
+
+1. Dynamic creation/registration of child devices on receipt of the very first data that they send.
+   This is desired at least for immediate child devices, if not for further nested child devices.
+1. Easy to create static routing rules so that it is easy to map a local MQTT topic for a nested child device
+   to the equivalent cloud topic for its twin.
+1. All the existing topics like `tedge/measurements`, `tedge/events` imply that the data received on these
+   must be forwarded to the cloud as well.
+   Currently there is no way to tell thin-edge to just route some data internally and not forward those to the cloud.
+   Since filtering and aggregation on the edge is a very common use-case, especially for local analytics, this is highly desired.
+1. It would be ideal if the context/source of data (tedge device, service or child device)
+   can be understood from the topic directly.
+   For e.g: a topic scheme like `tedge/main/{id}`, `tedge/service/{id}` and `tedge/child/{id}` is more user-friendly
+   than a simpler context agnostic scheme like `tedge/{id}` where `id` can be for any "thing".
+1. Allowing the "things" in a local domain to use their user-friendly unique IDs (within the tedge namespace)
+   in MQTT topics over their globally unique cloud IDs (typically very long and cumbersome) would be desired.
+   Ideally, all translation between user-provided-ids and cloud-twin-ids should be done internally by thin-edge,
+   as long as it doesn't adversely affect any data mapping cost at scale.
+
+### Out of scope
+
+1. Routing different kinds of data to different clouds, e.g: all telemetry to azure and all commands from/to Cumulocity.
+   Even though this requirement is realistic, thin-edge MQTT topics must not be corrupted with cloud specific semantics,
+   and the same requirement will be handled with some external routing mechanism(e.g: routing via bridge topics)
+1. Services on child devices. Service support can be limited to thin-edge devices only.
+
+## Proposals
+
+### Dedicated topics for tedge device, services and child devices
+
+The topics for the thin-edge device, the services running on it and child devices have different prefixes:
+
+For parent: tedge/main/<device-id>
+For services: tedge/service/<service-id>
+For immediate child devices: tedge/child/<child-id>
+For nested child devices: tedge/descendent/<child-id>
+
+#### Telemetry
+
+For telemetry data, the topics would be grouped under a `telemetry/` sub-topic as follows:
+* Measurements: `tedge/main/<device-id>/telemetry/measurements`
+* Events: `tedge/main/<device-id>/telemetry/events/<event-type>`
+* Alarms: `tedge/main/<device-id>/telemetry/alarms/<alarm-type>/<severity>`
+
+For child devices and services, a similar structure is followed like: `tedge/child/<child-id>/telemetry/measurements`,
+`tedge/service/<service-id>/telemetry/events/<event-type>` etc
+
+The subtopic levels `main/<device-id>` are really not required for the main device.
```

Comment:
This `<device-id>` in topic might come in handy, if there is a requirement to run multiple thin-edge instances on the same device (connected to multiple C8y instances, may be). I've heard such a use-case from @reubenmiller but would add that as a concrete requirement only once the use-case is clearer.

Reply:
Running multiple tedge instances on the same broker can be enabled without using the `<device-id>` in the topic by simply allowing users to also configure the root prefix, e.g. `tedge/`. For example users could configure `tedge1/` and `tedge2/` for two independent tedge installations which use the same broker. Each instance uses its own cloud bridge connection where the actual cloud device-id is used.

It preferred not to have the `<device-id>` in the topic name because it makes it harder to know what the topic name is, and the commands to publish mqtt messages on different devices then because device specific (which is makes for an annoying for users), and it also makes it more difficult to do cloud automation by sending a single mqtt pub commands to 1000 devices (e.g. sending a shell command via the Cumulocity `c8y_Command` operation to trigger a command or publish a message via the tedge MQTT topic, `tedge mqtt pub tedge/do_something '{}'`).